packages/beacon-node/src/chain/seenCache/seenBlockInput.ts

```diff
@@ -0,0 +1,253 @@
+import {RootHex, SignedBeaconBlock, Slot, deneb} from "@lodestar/types";
+import {LodestarError, Logger, toRootHex} from "@lodestar/utils";
+import {Metrics} from "../../metrics/metrics.js";
+import {IClock} from "../../util/clock.js";
+import {
+  BlockInputBlobs,
+  BlockInputPreData,
+  DAType,
+  ForkBlobsDA,
+  IBlockInput,
+  LogMetaBasic,
+  SourceMeta,
+  getDaOutOfRange,
+  isBlockInputBlobs,
+} from "../blocks/blockInput/index.js";
+import {ChainEvent, ChainEventEmitter} from "../emitter.js";
+import {BlobSidecarErrorCode, BlobSidecarGossipError} from "../errors/blobSidecarError.js";
+import {GossipAction} from "../errors/gossipValidation.js";
+
+export type SeenBlockInputCacheModules = {
+  config: ChainForkConfig;
+  clock: IClock;
+  chainEvents: ChainEventEmitter;
+  signal: AbortSignal;
+  //   custodyConfig: CustodyConfig;
+  metrics: Metrics | null;
+  logger?: Logger;
+};
+
+export type GetByBlobOptions = {
+  throwGossipErrorIfAlreadyKnown?: boolean;
+};
+
+/**
+ * Consumers that create BlockInputs or change types of old BlockInputs
+ *
+ * - gossipHandlers (block and blob)
+ * - beaconBlocksMaybeBlobsByRange
+ * - unavailableBeaconBlobsByRoot (beaconBlocksMaybeBlobsByRoot)
+ * - publishBlock in the beacon/blocks/index.ts API
+ *   https://github.com/ChainSafe/lodestar/blob/unstable/packages/beacon-node/src/api/impl/beacon/blocks/index.ts#L62
+ * - maybeValidateBlobs in verifyBlocksDataAvailability (is_data_available spec function)
+ *   https://github.com/ChainSafe/lodestar/blob/unstable/packages/beacon-node/src/chain/blocks/verifyBlocksDataAvailability.ts#L111
+ */
+
+export class SeenBlockInputCache {
+  private readonly config: ChainForkConfig;
+  private readonly clock: IClock;
+  private readonly chainEvents: ChainEventEmitter;
+  private readonly signal: AbortSignal;
+  private readonly metrics: Metrics | null;
+  private readonly logger?: Logger;
+  private blockInputs = new Map<RootHex, IBlockInput>();
+
+  constructor({config, clock, chainEvents, signal, metrics, logger}: SeenBlockInputCacheModules) {
+    this.config = config;
+    this.clock = clock;
+    this.chainEvents = chainEvents;
+    this.signal = signal;
+    this.metrics = metrics;
+    this.logger = logger;
+
+    if (metrics) {
+      metrics.seenCache.blockInput.blockInputCount.addCollect(() =>
+        metrics.seenCache.blockInput.blockInputCount.set(this.blockInputs.size)
+      );
+    }
+
+    this.chainEvents.on(ChainEvent.forkChoiceFinalized, this.onFinalized);
+    this.signal.addEventListener("abort", () => {
+      this.chainEvents.off(ChainEvent.forkChoiceFinalized, this.onFinalized);
+    });
+  }
+
+  has(rootHex: RootHex): boolean {
+    return this.blockInputs.has(rootHex);
+  }
+
+  get(rootHex: RootHex): IBlockInput | undefined {
+    return this.blockInputs.get(rootHex);
+  }
+
+  remove(rootHex: RootHex): void {
+    this.blockInputs.delete(rootHex);
+  }
+
+  prune(rootHex: RootHex): void {
```

Comment:
would be great to add a comment noting this does not only prune a block input but also all ancestors of it